File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        i=0
        j=1
        current_profit=0
        max_profit=0
        while i<j and j<len(prices):
            current_profit=prices[j]-prices[i]
            print(current_profit,prices[i],prices[j])
            if current_profit<0:
                current_profit=0
                i=j
            max_profit=max(max_profit,current_profit)
            j+=1
        return max_profit

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.java
// recursion 
class Solution {
    public int maxProfit(int[] p) {
        return helper(0,p,0,0,0);
        
    }

    public int helper(int i, int[] p, int currSum,int maxSum,int canSell){
        if(currSum>maxSum){
            maxSum=currSum;
        }
        if(i>=p.length){
            return maxSum;
        }
        if(canSell==0){
        return Math.max(helper(i+1,p,currSum-p[i],maxSum,1),
        helper(i+1,p,currSum,maxSum,0));
        }else{
        return    Math.max(helper(i+2,p,currSum+p[i],maxSum,0),
            helper(i+1,p,currSum,maxSum,1));
        }
    }
}
